frog_api/views/structure.py

```diff
@@ -0,0 +1,67 @@
+from typing import Any, List
+
+from django.db import models
+from frog_api.exceptions import *
+from frog_api.models import Category, Project
+from frog_api.serializers import ProjectSerializer
+from rest_framework import status
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from common import get_project, get_version
+
+
+class ProjectStructureView(APIView):
+    """
+    API endpoint that allows projects to be viewed.
+    """
+
+    def get(self, request: Request) -> Response:
+        """
+        Return a list of all projects.
+        """
+        projects = Project.objects.all()
+        serializer = ProjectSerializer(projects, many=True)
+        return Response(serializer.data)
+
+
+class CategoryStructureView(APIView):
+    """
+    API endpoint for adding new categories
+    """
+
+    @staticmethod
+    def create_categories(
+        req_data: dict[str, Any], project_slug: str, version_slug: str
+    ) -> List[Any]:
+        project = get_project(project_slug)
+        version = get_version(version_slug, project)
+
+        if "api_key" not in req_data:
+            raise MissingAPIKeyException()
+        if req_data["api_key"] != project.auth_key:
+            raise InvalidAPIKeyException()
+
+        categories = req_data["data"]
+
+        to_save: list[models.Model] = []
+        for cat in categories:
+            if Category.objects.filter(
+                slug=cat, version=version, version__project=project
+            ).exists():
+                raise AlreadyExistsException(
+                    f"Category {cat} already exists for project '{project_slug}', version '{version_slug}'"
+                )
+            to_save.append(Category(version=version, slug=cat, name=categories[cat]))
+
+        for s in to_save:
+            s.save()
+
+        return []
```

Comment:
Obviously we need to work out what to do here too.

Reply:
handled it the same as the other one for now :)